File: backend/neural/neuron.py
"""
Single Neuron Implementation for CAMBRIC LABS

A neuron is the fundamental unit of a neural network. It receives inputs,
multiplies each by a weight, adds them up, adds a bias, and applies an
activation function to produce an output.

Mathematical Model:
    z = Σ(xᵢ × wᵢ) + b      (weighted sum)
    y = activation(z)        (output)

Forward: compute output
Backward: compute gradients (NO weight changes)
Update: apply gradients to weights

This separation allows CAMBRIC Labs to:
1. Inspect gradients before they're applied
2. Support multiple optimizers
3. Visualize the learning process
"""

import numpy as np
from typing import List, Dict, Any, Optional
from dataclasses import dataclass


@dataclass
class GradientResult:
    """Result of backward pass - contains all computed gradients."""
    input_gradients: np.ndarray      # dL/dx for each input
    weight_gradients: np.ndarray    # dL/dw for each weight
    bias_gradient: float            # dL/db
    activation_gradient: float      # dL/dz (before activation)
    weighted_sum: float             # z value (before activation)
    output: float                   # activation(z)


class Neuron:
    """
    A single artificial neuron with separated forward/backward/update.
    
    The neuron computes: output = activation(sum(inputs * weights) + bias)
    
    Key Design Principles:
    - backward() computes gradients but does NOT modify weights
    - update() applies gradients to weights
    - reset_cache() clears temporary data, preserves weights
    - reset_parameters() reinitializes weights
    
    Attributes:
        input_count: Number of inputs this neuron expects
        weights: Trainable weight values
        bias: Trainable bias value
        activation: Name of the activation function
    """
    
    SUPPORTED_ACTIVATIONS = ['relu', 'sigmoid', 'tanh', 'identity', 'leaky_relu']
    VERSION = "2.0.0"
    
    def __init__(
        self,
        input_count: int,
        weights: Optional[List[float]] = None,
        bias: float = 0.0,
        activation: str = 'relu',
        seed: Optional[int] = None
    ):
        """
        Initialize a neuron.
        
        Args:
            input_count: Number of input connections
            weights: Initial weights (random if None)
            bias: Initial bias value
            activation: Activation function name
            seed: Random seed for reproducibility
        """
        if activation not in self.SUPPORTED_ACTIVATIONS:
            raise ValueError(
                f"Unknown activation: {activation}. "
                f"Choose from: {self.SUPPORTED_ACTIVATIONS}"
            )
        
        if input_count < 1:
            raise ValueError(f"input_count must be >= 1, got {input_count}")
        
        self.input_count = input_count
        self.activation = activation
        
        # Use local random generator for reproducibility
        rng = np.random.default_rng(seed)
        
        if weights is None:
            # Activation-appropriate initialization
            self.weights = self._initialize_weights(rng)
        else:
            if len(weights) != input_count:
                raise ValueError(
                    f"Weight count ({len(weights)}) doesn't match "
                    f"input_count ({input_count})"
                )
            self.weights = np.array(weights, dtype=np.float64)
        
        self.bias = float(bias)
        
        # Cache for backward pass - these are TEMPORARY
        self._last_input: Optional[np.ndarray] = None
        self._last_weighted_sum: Optional[float] = None
        self._last_output: Optional[float] = None
        self._last_gradients: Optional[GradientResult] = None
        
        # Statistics for educational purposes
        self._training_stats = {
            'forward_calls': 0,
            'backward_calls': 0,
            'update_calls': 0
        }
    
    def _initialize_weights(self, rng: np.random.Generator) -> np.ndarray:
        """
        Initialize weights based on activation function.
        
        He initialization for ReLU variants: std = sqrt(2 / input_count)
        Xavier initialization for sigmoid/tanh: std = sqrt(1 / input_count)
        """
        if self.activation == 'relu':
            # He initialization
            std = np.sqrt(2.0 / self.input_count)
        elif self.activation in ('sigmoid', 'tanh'):
            # Xavier initialization
            std = np.sqrt(1.0 / self.input_count)
        else:
            # Default: Xavier
            std = np.sqrt(2.0 / self.input_count)
        
        return rng.normal(0, std, self.input_count).astype(np.float64)
    
    @property
    def parameter_count(self) -> int:
        """Number of trainable parameters (weights + bias)."""
        return self.input_count + 1
    
    def forward(self, inputs: List[float]) -> Dict[str, Any]:
        """
        Compute the forward pass through this neuron.
        
        Args:
            inputs: List of input values
            
        Returns:
            Dictionary containing forward pass details for visualization.
        """
        # Validate inputs
        inputs_arr = np.array(inputs, dtype=np.float64)
        
        if len(inputs_arr) != self.input_count:
            raise ValueError(
                f"Input count ({len(inputs_arr)}) doesn't match "
                f"neuron's input_count ({self.input_count})"
            )
        
        if np.any(np.isnan(inputs_arr)) or np.any(np.isinf(inputs_arr)):
            raise ValueError("Input contains NaN or Infinity values")
        
        self._last_input = inputs_arr.copy()
        
        # Compute weighted sum: z = Σ(xᵢ × wᵢ) + b
        contributions = inputs_arr * self.weights
        weighted_sum = np.sum(contributions) + self.bias
        self._last_weighted_sum = float(weighted_sum)
        
        # Apply activation: y = activation(z)
        output = self._apply_activation(weighted_sum)
        self._last_output = float(output)
        
        self._training_stats['forward_calls'] += 1
        
        return {
            'output': float(output),
            'weighted_sum': float(weighted_sum),
            'contributions': contributions.tolist(),
            'activation_used': self.activation,
            'inputs': inputs_arr.tolist(),
            'weights': self.weights.tolist(),
            'bias': float(self.bias)
        }
    
    def _apply_activation(self, x: float) -> float:
        """Apply the activation function."""
        if self.activation == 'relu':
            return max(0.0, x)
        elif self.activation == 'leaky_relu':
            return x if x > 0 else 0.01 * x
        elif self.activation == 'sigmoid':
            # Numerically stable sigmoid
            if x < -500:
                return 0.0
            elif x > 500:
                return 1.0
            return 1.0 / (1.0 + np.exp(-x))
        elif self.activation == 'tanh':
            return float(np.tanh(x))
        elif self.activation == 'identity':
            return x
        else:
            return x  # Default to identity
    
    def _get_activation_derivative(self, output: float, weighted_sum: float) -> float:
        """Get the derivative of the activation function."""
        if self.activation == 'relu':
            # For ReLU, we need the input to the activation (weighted_sum)
            return 1.0 if weighted_sum > 0 else 0.0
        elif self.activation == 'leaky_relu':
            return 1.0 if weighted_sum > 0 else 0.01
        elif self.activation == 'sigmoid':
            # σ(x) * (1 - σ(x)) - using output which is σ(x)
            return output * (1.0 - output)
        elif self.activation == 'tanh':
            # 1 - tanh²(x) - using output which is tanh(x)
            return 1.0 - output * output
        elif self.activation == 'identity':
            return 1.0
        else:
            return 1.0
    
    def backward(self, output_gradient: float) -> GradientResult:
        """
        Compute gradients via backpropagation.
        
        IMPORTANT: This method ONLY computes gradients.
        It does NOT modify any parameters.
        
        Mathematical derivation:
        - dL/dz = dL/dy × dy/dz  (chain rule)
        - dL/dwᵢ = dL/dz × xᵢ   (for each weight)
        - dL/db = dL/dz           (for bias)
        - dL/dxᵢ = dL/dz × wᵢ    (for each input)
        
        Args:
            output_gradient: Gradient of loss with respect to this neuron's output (dL/dy)
            
        Returns:
            GradientResult containing all computed gradients.
            Use update() to apply these gradients.
        """
        if self._last_input is None:
            raise RuntimeError(
                "Must call forward() before backward(). "
                "The neuron needs cached values from forward pass."
            )
        
        if np.isnan(output_gradient) or np.isinf(output_gradient):
            raise ValueError(f"Invalid output_gradient: {output_gradient}")
        
        # Activation derivative: dy/dz
        act_deriv = self._get_activation_derivative(
            self._last_output, 
            self._last_weighted_sum
        )
        
        # Chain rule: dL/dz = dL/dy × dy/dz
        activation_gradient = output_gradient * act_deriv
        
        # Weight gradients: dL/dwᵢ = dL/dz × xᵢ
        weight_gradients = self._last_input * activation_gradient
        
        # Bias gradient: dL/db = dL/dz
        bias_gradient = float(activation_gradient)
        
        # Input gradients (for previous layer): dL/dxᵢ = dL/dz × wᵢ
        input_gradients = self.weights * activation_gradient
        
        # Store for inspection
        self._last_gradients = GradientResult(
            input_gradients=input_gradients.copy(),
            weight_gradients=weight_gradients.copy(),
            bias_gradient=bias_gradient,
            activation_gradient=float(activation_gradient),
            weighted_sum=float(self._last_weighted_sum),
            output=float(self._last_output)
        )
        
        self._training_stats['backward_calls'] += 1
        
        return self._last_gradients
    
    def get_gradients(self) -> Optional[GradientResult]:
        """Get the last computed gradients (for inspection)."""
        return self._last_gradients
    
    def update(
        self, 
        gradients: GradientResult, 
        learning_rate: float,
        clip_gradients: Optional[float] = None
    ) -> Dict[str, Any]:
        """
        Apply gradients to update parameters.
        
        Uses gradient descent: θ_new = θ_old - lr × gradient
        
        Args:
            gradients: Gradients computed by backward()
            learning_rate: Learning rate (step size)
            clip_gradients: Optional max gradient norm for clipping
            
        Returns:
            Dictionary with update details.
        """
        if learning_rate <= 0:
            raise ValueError(f"learning_rate must be positive, got {learning_rate}")
        
        if learning_rate > 1:
            import warnings
            warnings.warn(
                f"Large learning_rate ({learning_rate}) may cause unstable training",
                UserWarning
            )
        
        # Get gradients to apply
        w_grads = gradients.weight_gradients.copy()
        b_grad = gradients.bias_gradient
        
        # Optional gradient clipping
        if clip_gradients is not None and clip_gradients > 0:
            w_norm = np.linalg.norm(w_grads)
            if w_norm > clip_gradients:
                w_grads = w_grads * (clip_gradients / w_norm)
                b_grad = b_grad * (clip_gradients / w_norm)
        
        # Store old values for comparison
        old_weights = self.weights.copy()
        old_bias = self.bias
        
        # Apply updates: parameter_new = parameter_old - lr × gradient
        self.weights -= learning_rate * w_grads
        self.bias -= learning_rate * b_grad
        
        self._training_stats['update_calls'] += 1
        
        return {
            'weight_changes': (-learning_rate * w_grads).tolist(),
            'bias_change': -learning_rate * b_grad,
            'old_weights': old_weights.tolist(),
            'new_weights': self.weights.tolist(),
            'old_bias': float(old_bias),
            'new_bias': float(self.bias),
            'gradient_norms': {
                'weight_norm': float(np.linalg.norm(w_grads)),
                'bias_gradient': float(b_grad)
            }
        }
    
    def reset_cache(self) -> None:
        """
        Reset temporary cache values.
        
        Preserves learned parameters (weights, bias).
        Call this before a new forward pass.
        """
        self._last_input = None
        self._last_weighted_sum = None
        self._last_output = None
        self._last_gradients = None
    
    def reset_parameters(self, seed: Optional[int] = None) -> None:
        """
        Reset learned parameters to new random values.
        
        Use this to reinitialize training.
        """
        rng = np.random.default_rng(seed)
        self.weights = self._initialize_weights(rng)
        self.bias = 0.0
        self._training_stats = {
            'forward_calls': 0,
            'backward_calls': 0,
            'update_calls': 0
        }
    
    def reset(self) -> None:
        """
        Reset internal state.
        
        Alias for reset_cache() for backwards compatibility.
        """
        self.reset_cache()
    
    def get_state(self) -> Dict[str, Any]:
        """Get current state for serialization."""
        return {
            'type': 'neuron',
            'version': self.VERSION,
            'input_count': self.input_count,
            'activation': self.activation,
            'weights': self.weights.tolist(),
            'bias': float(self.bias),
            'parameter_count': self.parameter_count,
            'training_stats': self._training_stats.copy()
        }
    
    def set_weights(self, weights: List[float]) -> None:
        """Set new weights."""
        if len(weights) != self.input_count:
            raise ValueError(
                f"Weight count ({len(weights)}) doesn't match "
                f"input_count ({self.input_count})"
            )
        self.weights = np.array(weights, dtype=np.float64)
    
    def set_bias(self, bias: float) -> None:
        """Set new bias."""
        self.bias = float(bias)
    
    def inspect(self) -> Dict[str, Any]:
        """
        Get detailed state for educational visualization.
        
        Returns comprehensive information for CAMBRIC Labs UI.
        """
        state = self.get_state()
        
        if self._last_input is not None:
            state['cached'] = {
                'inputs': self._last_input.tolist(),
                'weighted_sum': self._last_weighted_sum,
                'output': self._last_output
            }
        
        if self._last_gradients is not None:
            g = self._last_gradients
            state['gradients'] = {
                'input_gradients': g.input_gradients.tolist(),
                'weight_gradients': g.weight_gradients.tolist(),
                'bias_gradient': g.bias_gradient,
                'activation_gradient': g.activation_gradient,
                'weighted_sum': g.weighted_sum
            }
        
        return state
    
    def __repr__(self) -> str:
        return (
            f"Neuron(input_count={self.input_count}, "
            f"activation={self.activation}, "
            f"params={self.parameter_count})"
        )


def create_neuron(
    input_count: int,
    activation: str = 'relu',
    seed: Optional[int] = None
) -> Neuron:
    """
    Factory function to create a neuron with default initialization.
    
    Args:
        input_count: Number of inputs
        activation: Activation function name
        seed: Random seed for reproducibility
        
    Returns:
        New Neuron instance
    """
    return Neuron(
        input_count=input_count,
        activation=activation,
        seed=seed
    )


def numerical_gradient_check(
    neuron: Neuron,
    inputs: List[float],
    target: float,
    epsilon: float = 1e-5
) -> Dict[str, Any]:
    """
    Verify gradients using finite differences.
    
    For each parameter θ:
    numerical_gradient ≈ (loss(θ + ε) - loss(θ - ε)) / (2ε)
    
    Args:
        neuron: Neuron to check
        inputs: Input values
        target: Target output
        epsilon: Small perturbation
        
    Returns:
        Comparison of analytical vs numerical gradients
    """
    # Forward pass
    result = neuron.forward(inputs)
    output = result['output']
    
    # Simple MSE loss
    loss_plus = (output - target) ** 2
    loss_minus = (output - target) ** 2
    
    # Store original weights
    orig_weights = neuron.weights.copy()
    orig_bias = neuron.bias
    
    # Check weight gradients
    weight_grads_analytical = []
    weight_grads_numerical = []
    
    for i in range(neuron.input_count):
        # loss(w + ε)
        neuron.weights = orig_weights.copy()
        neuron.weights[i] += epsilon
        out_plus = neuron.forward(inputs)['output']
        loss_plus = (out_plus - target) ** 2
        
        # loss(w - ε)
        neuron.weights = orig_weights.copy()
        neuron.weights[i] -= epsilon
        out_minus = neuron.forward(inputs)['output']
        loss_minus = (out_minus - target) ** 2
        
        # Numerical gradient
        numerical = (loss_plus - loss_minus) / (2 * epsilon)
        weight_grads_numerical.append(numerical)
        
        neuron.weights = orig_weights.copy()
    
    # Restore original
    neuron.weights = orig_weights
    neuron.bias = orig_bias
    
    return {
        'analytical_weight_gradients': weight_grads_analytical if 'weight_grads_analytical' in dir() else [],
        'numerical_weight_gradients': weight_grads_numerical,
        'epsilon': epsilon,
        'passed': True  # Would need actual analytical to compare
    }

File: frontend/src/engine/neuron.ts
/**
 * Single Neuron Implementation for CAMBRIC LABS
 *
 * Ported 1:1 from backend/neural/neuron.py.
 *
 * z = Σ(xᵢ × wᵢ) + b      (weighted sum)
 * y = activation(z)        (output)
 *
 * forward()  -> compute output
 * backward() -> compute gradients (does NOT change weights)
 * update()   -> apply gradients to weights
 *
 * This separation lets the UI inspect gradients before they're applied,
 * support multiple optimizers later, and visualize the learning process.
 */

import { ActivationName } from './activation';

export type { ActivationName };

export const SUPPORTED_ACTIVATIONS: ActivationName[] = [
  'relu',
  'sigmoid',
  'tanh',
  'identity',
  'leaky_relu',
];

export interface GradientResult {
  inputGradients: number[]; // dL/dx for each input
  weightGradients: number[]; // dL/dw for each weight
  biasGradient: number; // dL/db
  activationGradient: number; // dL/dz (before activation)
  weightedSum: number; // z value (before activation)
  output: number; // activation(z)
}

export interface ForwardResult {
  output: number;
  weightedSum: number;
  contributions: number[];
  activationUsed: ActivationName;
  inputs: number[];
  weights: number[];
  bias: number;
}

export interface UpdateResult {
  weightChanges: number[];
  biasChange: number;
  oldWeights: number[];
  newWeights: number[];
  oldBias: number;
  newBias: number;
  gradientNorms: { weightNorm: number; biasGradient: number };
}

export interface NeuronState {
  type: 'neuron';
  version: string;
  inputCount: number;
  activation: ActivationName;
  weights: number[];
  bias: number;
  parameterCount: number;
  trainingStats: { forwardCalls: number; backwardCalls: number; updateCalls: number };
}

/** Simple seeded PRNG (mulberry32) + Box-Muller normal sampler, mirroring numpy's default_rng usage. */
class SeededRandom {
  private state: number;

  constructor(seed?: number) {
    this.state = seed === undefined ? (Math.random() * 2 ** 32) >>> 0 : seed >>> 0;
  }

  private next(): number {
    this.state |= 0;
    this.state = (this.state + 0x6d2b79f5) | 0;
    let t = Math.imul(this.state ^ (this.state >>> 15), 1 | this.state);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  }

  /** Sample from N(mean, std) via Box-Muller. */
  normal(mean: number, std: number): number {
    let u = 0;
    let v = 0;
    while (u === 0) u = this.next();
    while (v === 0) v = this.next();
    const z = Math.sqrt(-2 * Math.log(u)) * Math.cos(2 * Math.PI * v);
    return mean + z * std;
  }
}

function dot(a: number[], b: number[]): number[] {
  return a.map((v, i) => v * b[i]);
}

function norm(a: number[]): number {
  return Math.sqrt(a.reduce((sum, v) => sum + v * v, 0));
}

export class Neuron {
  static readonly VERSION = '2.0.0';

  readonly inputCount: number;
  activation: ActivationName;
  weights: number[];
  bias: number;

  private lastInput: number[] | null = null;
  private lastWeightedSum: number | null = null;
  private lastOutput: number | null = null;
  private lastGradients: GradientResult | null = null;

  private trainingStats = { forwardCalls: 0, backwardCalls: 0, updateCalls: 0 };

  constructor(opts: {
    inputCount: number;
    weights?: number[];
    bias?: number;
    activation?: ActivationName;
    seed?: number;
  }) {
    const { inputCount, weights, bias = 0.0, activation = 'relu', seed } = opts;

    if (!SUPPORTED_ACTIVATIONS.includes(activation)) {
      throw new Error(
        `Unknown activation: ${activation}. Choose from: ${SUPPORTED_ACTIVATIONS.join(', ')}`
      );
    }
    if (inputCount < 1) {
      throw new Error(`inputCount must be >= 1, got ${inputCount}`);
    }

    this.inputCount = inputCount;
    this.activation = activation;

    const rng = new SeededRandom(seed);

    if (weights === undefined) {
      this.weights = this.initializeWeights(rng);
    } else {
      if (weights.length !== inputCount) {
        throw new Error(
          `Weight count (${weights.length}) doesn't match inputCount (${inputCount})`
        );
      }
      this.weights = [...weights];
    }

    this.bias = bias;
  }

  /** He init for ReLU variants (std = sqrt(2/n)); Xavier for sigmoid/tanh (std = sqrt(1/n)). */
  private initializeWeights(rng: SeededRandom): number[] {
    let std: number;
    if (this.activation === 'relu') {
      std = Math.sqrt(2.0 / this.inputCount);
    } else if (this.activation === 'sigmoid' || this.activation === 'tanh') {
      std = Math.sqrt(1.0 / this.inputCount);
    } else {
      std = Math.sqrt(2.0 / this.inputCount);
    }
    return Array.from({ length: this.inputCount }, () => rng.normal(0, std));
  }

  get parameterCount(): number {
    return this.inputCount + 1;
  }

  forward(inputs: number[]): ForwardResult {
    if (inputs.length !== this.inputCount) {
      throw new Error(
        `Input count (${inputs.length}) doesn't match neuron's inputCount (${this.inputCount})`
      );
    }
    if (inputs.some((v) => Number.isNaN(v) || !Number.isFinite(v))) {
      throw new Error('Input contains NaN or Infinity values');
    }

    this.lastInput = [...inputs];

    const contributions = dot(inputs, this.weights);
    const weightedSum = contributions.reduce((a, b) => a + b, 0) + this.bias;
    this.lastWeightedSum = weightedSum;

    const output = this.applyActivation(weightedSum);
    this.lastOutput = output;

    this.trainingStats.forwardCalls += 1;

    return {
      output,
      weightedSum,
      contributions,
      activationUsed: this.activation,
      inputs: [...inputs],
      weights: [...this.weights],
      bias: this.bias,
    };
  }

  private applyActivation(x: number): number {
    switch (this.activation) {
      case 'relu':
        return Math.max(0, x);
      case 'leaky_relu':
        return x > 0 ? x : 0.01 * x;
      case 'sigmoid': {
        if (x < -500) return 0.0;
        if (x > 500) return 1.0;
        return 1.0 / (1.0 + Math.exp(-x));
      }
      case 'tanh':
        return Math.tanh(x);
      case 'identity':
        return x;
      default:
        return x;
    }
  }

  private activationDerivative(output: number, weightedSum: number): number {
    switch (this.activation) {
      case 'relu':
        return weightedSum > 0 ? 1.0 : 0.0;
      case 'leaky_relu':
        return weightedSum > 0 ? 1.0 : 0.01;
      case 'sigmoid':
        return output * (1.0 - output);
      case 'tanh':
        return 1.0 - output * output;
      case 'identity':
        return 1.0;
      default:
        return 1.0;
    }
  }

  /**
   * Backpropagation. ONLY computes gradients — does NOT modify parameters.
   *
   * dL/dz = dL/dy × dy/dz
   * dL/dwᵢ = dL/dz × xᵢ
   * dL/db  = dL/dz
   * dL/dxᵢ = dL/dz × wᵢ
   */
  backward(outputGradient: number): GradientResult {
    if (this.lastInput === null || this.lastOutput === null || this.lastWeightedSum === null) {
      throw new Error('Must call forward() before backward(). The neuron needs cached values.');
    }
    if (Number.isNaN(outputGradient) || !Number.isFinite(outputGradient)) {
      throw new Error(`Invalid outputGradient: ${outputGradient}`);
    }

    const actDeriv = this.activationDerivative(this.lastOutput, this.lastWeightedSum);
    const activationGradient = outputGradient * actDeriv;

    const weightGradients = this.lastInput.map((x) => x * activationGradient);
    const biasGradient = activationGradient;
    const inputGradients = this.weights.map((w) => w * activationGradient);

    this.lastGradients = {
      inputGradients,
      weightGradients,
      biasGradient,
      activationGradient,
      weightedSum: this.lastWeightedSum,
      output: this.lastOutput,
    };

    this.trainingStats.backwardCalls += 1;

    return this.lastGradients;
  }

  getGradients(): GradientResult | null {
    return this.lastGradients;
  }

  /** Applies gradients via gradient descent: θ_new = θ_old - lr × gradient. */
  update(gradients: GradientResult, learningRate: number, clipGradients?: number): UpdateResult {
    if (learningRate <= 0) {
      throw new Error(`learningRate must be positive, got ${learningRate}`);
    }
    if (learningRate > 1) {
      // eslint-disable-next-line no-console
      console.warn(`Large learningRate (${learningRate}) may cause unstable training`);
    }

    let wGrads = [...gradients.weightGradients];
    let bGrad = gradients.biasGradient;

    if (clipGradients !== undefined && clipGradients > 0) {
      const wNorm = norm(wGrads);
      if (wNorm > clipGradients) {
        const scale = clipGradients / wNorm;
        wGrads = wGrads.map((w) => w * scale);
        bGrad *= scale;
      }
    }

    const oldWeights = [...this.weights];
    const oldBias = this.bias;

    this.weights = this.weights.map((w, i) => w - learningRate * wGrads[i]);
    this.bias -= learningRate * bGrad;

    this.trainingStats.updateCalls += 1;

    return {
      weightChanges: wGrads.map((g) => -learningRate * g),
      biasChange: -learningRate * bGrad,
      oldWeights,
      newWeights: [...this.weights],
      oldBias,
      newBias: this.bias,
      gradientNorms: { weightNorm: norm(wGrads), biasGradient: bGrad },
    };
  }

  /** Clears temporary cache; preserves learned weights/bias. Call before a new forward pass. */
  resetCache(): void {
    this.lastInput = null;
    this.lastWeightedSum = null;
    this.lastOutput = null;
    this.lastGradients = null;
  }

  /** Reinitializes weights/bias to new random values. */
  resetParameters(seed?: number): void {
    const rng = new SeededRandom(seed);
    this.weights = this.initializeWeights(rng);
    this.bias = 0.0;
    this.trainingStats = { forwardCalls: 0, backwardCalls: 0, updateCalls: 0 };
  }

  /** Alias for resetCache(), kept for parity with the Python API. */
  reset(): void {
    this.resetCache();
  }

  getState(): NeuronState {
    return {
      type: 'neuron',
      version: Neuron.VERSION,
      inputCount: this.inputCount,
      activation: this.activation,
      weights: [...this.weights],
      bias: this.bias,
      parameterCount: this.parameterCount,
      trainingStats: { ...this.trainingStats },
    };
  }

  setWeights(weights: number[]): void {
    if (weights.length !== this.inputCount) {
      throw new Error(
        `Weight count (${weights.length}) doesn't match inputCount (${this.inputCount})`
      );
    }
    this.weights = [...weights];
  }

  setBias(bias: number): void {
    this.bias = bias;
  }

  /** Detailed state (including cached forward/backward values) for UI visualization. */
  inspect(): NeuronState & {
    cached?: { inputs: number[]; weightedSum: number; output: number };
    gradients?: {
      inputGradients: number[];
      weightGradients: number[];
      biasGradient: number;
      activationGradient: number;
      weightedSum: number;
    };
  } {
    const state = this.getState() as ReturnType<Neuron['inspect']>;

    if (this.lastInput !== null && this.lastWeightedSum !== null && this.lastOutput !== null) {
      state.cached = {
        inputs: [...this.lastInput],
        weightedSum: this.lastWeightedSum,
        output: this.lastOutput,
      };
    }

    if (this.lastGradients !== null) {
      const g = this.lastGradients;
      state.gradients = {
        inputGradients: [...g.inputGradients],
        weightGradients: [...g.weightGradients],
        biasGradient: g.biasGradient,
        activationGradient: g.activationGradient,
        weightedSum: g.weightedSum,
      };
    }

    return state;
  }
}

/** Factory function mirroring create_neuron() in the Python engine. */
export function createNeuron(
  inputCount: number,
  activation: ActivationName = 'relu',
  seed?: number
): Neuron {
  return new Neuron({ inputCount, activation, seed });
}

/**
 * Verify gradients using finite differences (numerical gradient checking).
 * numerical_gradient ≈ (loss(θ+ε) - loss(θ-ε)) / (2ε)
 *
 * Note: this version computes BOTH the analytical and numerical weight
 * gradients and compares them directly — the original Python utility had
 * a bug where the analytical side was never actually populated.
 */
export function numericalGradientCheck(
  neuron: Neuron,
  inputs: number[],
  target: number,
  epsilon = 1e-5
): {
  analyticalWeightGradients: number[];
  numericalWeightGradients: number[];
  epsilon: number;
  maxAbsoluteDifference: number;
  passed: boolean;
} {
  const forwardResult = neuron.forward(inputs);
  const outputGradient = 2 * (forwardResult.output - target); // d(MSE)/d(output)
  const analytical = neuron.backward(outputGradient);
  const analyticalWeightGradients = [...analytical.weightGradients];

  const origWeights = [...neuron.weights];
  const origBias = neuron.bias;

  const numericalWeightGradients: number[] = [];

  for (let i = 0; i < neuron.inputCount; i++) {
    neuron.setWeights(origWeights);
    const wPlus = [...origWeights];
    wPlus[i] += epsilon;
    neuron.setWeights(wPlus);
    const outPlus = neuron.forward(inputs).output;
    const lossPlus = (outPlus - target) ** 2;

    const wMinus = [...origWeights];
    wMinus[i] -= epsilon;
    neuron.setWeights(wMinus);
    const outMinus = neuron.forward(inputs).output;
    const lossMinus = (outMinus - target) ** 2;

    numericalWeightGradients.push((lossPlus - lossMinus) / (2 * epsilon));
  }

  neuron.setWeights(origWeights);
  neuron.setBias(origBias);

  const maxAbsoluteDifference = Math.max(
    ...analyticalWeightGradients.map((a, i) => Math.abs(a - numericalWeightGradients[i]))
  );

  return {
    analyticalWeightGradients,
    numericalWeightGradients,
    epsilon,
    maxAbsoluteDifference,
    passed: maxAbsoluteDifference < 1e-3,
  };
}
